compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -138,6 +313,93 @@ emitLinkedTuningSpec(ModuleOp module, ArrayRef<NamedSequenceOp> specsToLink) {
   return newSpec;
 }
 
+static FailureOr<NamedSequenceOp> emitLinkedDefaultTuningSpec(ModuleOp module) {
+  OpBuilder builder(module.getContext());
+
+  // Step 1: Collect NamedSequenceOps from inner modules.
+  TuningSpecsToMerge tuningSpecs = collectTuningSpecsToMerge(module);
+  SmallVector<NamedSequenceOp> &namedSequenceOpsToMove =
+      tuningSpecs.namedSequenceOpsToMove;
+  llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> &namedSequenceToForeachMatch =
+      tuningSpecs.namedSequenceToForeachMatch;
+
+  // Step 2-a: Make sure the name sequence names are unique, and then move
+  // collected NamedSequenceOps to the top-level module.
+  llvm::StringMap<unsigned> specNameCounts;
+  for (NamedSequenceOp op : namedSequenceOpsToMove) {
+    updateNamedSequenceOp(op, module, builder, specNameCounts,
+                          namedSequenceToForeachMatch);
+  }
+
+  // Step 2-b: Create a new NamedSequenceOp `__kernel_config` in the top-level
+  // module.
+
+  // Collect all `ForeachMatchOp`s from `__kernel_config` NamedSequenceOps
+  // in inner modules to merge into the new `__kernel_config` in the top module.
+  SmallVector<ForeachMatchOp> foreachMatchOps;
+  for (auto innerModule : module.getBody()->getOps<ModuleOp>()) {
+    auto namedSequenceOp = *innerModule.getOps<NamedSequenceOp>().begin();
+    assert(
+        namedSequenceOp.getSymName() == kKernelConfigSpecName &&
+        namedSequenceOp->hasAttr(kTuningSpecEntrypointAttrName) &&
+        "NamedSequenceOp must have the expected symbol name `__kernel_config` "
+        "and attribute `iree_codegen.tuning_spec_entrypoint`.");
+    ForeachMatchOp foreachMatch =
+        *namedSequenceOp.getOps<ForeachMatchOp>().begin();
+    foreachMatchOps.push_back(foreachMatch);
+  }
+
+  builder.setInsertionPointToEnd(module.getBody());
+  Location loc = module.getLoc();
+  auto newNamedSequence =
```

Comment:
Write the full type here since it's not obvious based on the context: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable